Document Actions: Fix document title misalignment with an open nav sidebar

jeyip opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
The document title in the site editor isn't center aligned with content when the new navigation sidebar is open.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Set up the site editor in a local dev environment
2. Navigate to the site editor from the wp-admin sidebar
3. Note the centered positioning of the topbar document title
4. Open the new navigation sidebar by clicking on the wordpress icon in the top left corner of the screen
5. Note that the positioning of the topbar document title is still centered with editor content

## Screenshots <!-- if applicable -->
### Before
![Before](https://user-images.githubusercontent.com/5414230/94202703-fc3ffc00-fe72-11ea-90b8-10e7f6e24cbb.gif)

### After
![After](https://user-images.githubusercontent.com/5414230/94202592-cc90f400-fe72-11ea-984f-c9ef1783e79f.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix which addresses https://github.com/WordPress/gutenberg/issues/25629.

Elaborating more on the approach, there are two scenarios we have to support.

1. When the nav sidebar is closed, we want to center the document title with editor content, which happens to match the browser viewport.
2. When the nav sidebar is open, the center of editor content no longer matches the center of the browser viewport. The center becomes a segment of the viewport. 

Moving the nav sidebar component outside of the `edit-site-header_start` wrapper with the existing flex layout solves # 2. To handle situation # 1, I absolutely position the document title in the middle of the page. Opening the sidebar just removes all references to absolute positioning. Closing the sidebar reintroduces absolute positioning.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

jeyip committed:
Fix header misalignment

jeyip committed:
Rename layout classes to be position independent

Addison-Stavlo reviewed (commented):
This fixes it, but I'm a little confused regarding the extra wrappers, classnames, and added style rules in general.  Simply moving the `MainDashboardButton.Slot` out of the `edit-site-header_start` wrapper solves the issue on my end.

noahtallen commented:
i thought that approach might fix it too, but since the side button is no longer part of `edit-site-header_start`, it is centered based on the rest of the elements. so it's a little bit off center:

<img width="1924" alt="Screen Shot 2020-09-24 at 5 45 13 PM" src="https://user-images.githubusercontent.com/6265975/94214302-0fac9080-fe8e-11ea-9486-21fbe5cb54ab.png">

(I think that could be fixed by adding absolute and some padding)

Addison-Stavlo commented:
> it is centered based on the rest of the elements. so it's a little bit off center

Ahhh yeah! I see that now.

jeyip commented:
> This fixes it, but I'm a little confused regarding the extra wrappers, classnames, and added style rules in general. 

I updated the classes because they were previously based on positioning (start, center, end). With this fix, `.edit-site-header_start`, wouldn't be the start of the header anymore. It would appear _after_ the sidebar toggle.

I don't feel strongly about the changes though, so if we feel like they're unintuitive, I'm happy to revert them. @Addison-Stavlo

Addison-Stavlo commented:
> so if we feel like they're unintuitive, I'm happy to revert them.

Actually regarding the name changes themselves they make sense to me!  I was just confused why there were extra wrappers and added style rules when moving that sidebar out of the header seemed to solve the issue (which it really didn't as Noah pointed out).

jeyip commented:
I added a few notes to the PR subheading "Types of Changes" to elaborate more on what the additional styling is meant to do

noahtallen commented:
> When the nav sidebar is closed, we want to center the document title in the browser viewport.
> When the nav sidebar is open, we want to center the document title with editor content.

A nitpicky point on this is that we're really trying to align with editor content all the time (I think). It just so happens that it matches the viewport when it's fullscreen :)

david-szabo97 commented:
> A nitpicky point on this is that we're really trying to align with editor content all the time (I think). It just so happens that it matches the viewport when it's fullscreen :)

If we have a scrollbar in the content then the document title is a little bit off. Since the editor's content container's width is smaller with a few pixels because of the scrollbar.

Without scrollbar
![image](https://user-images.githubusercontent.com/2256104/94249112-b725d380-ff1f-11ea-8a5c-0fcff842597b.png)

With scrollbar
![image](https://user-images.githubusercontent.com/2256104/94249149-c442c280-ff1f-11ea-9d12-be2b5883d547.png)

Addison-Stavlo reviewed (commented):
One thing I am noticing is the difference in overlapping behavior on smaller screen widths / devices.

With these changes, when we reduce the screen width the items on the left overlap our name panel:
![new-overlap-sized](https://user-images.githubusercontent.com/28742426/94285810-30262a80-ff22-11ea-9b9d-efccbecab7c3.gif)

On the current master branch, this is prevented given how the style rules are set:
![current-non-overlap](https://user-images.githubusercontent.com/28742426/94285851-4338fa80-ff22-11ea-946b-39e65d980997.gif)

It would be nice to preserve this behavior, however it sounds like the double-dropdown that it is conflicting with will be removed in the future and I don't think there is an issue with the remainder of the items on the left (unless more are added in the future) as the top toolbar will disappear before we get to that narrow of a screen width.  Given that I am not sure how concerning this should be. 🤔

jeyip commented:
> If we have a scrollbar in the content then the document title is a little bit off. Since the editor's content container's width is smaller with a few pixels because of the scrollbar.

> It would be nice to preserve this behavior, however it sounds like the double-dropdown that it is conflicting with will be removed in the future and I don't think there is an issue with the remainder of the items on the left (unless more are added in the future)

@david-szabo97 @Addison-Stavlo Great calllouts. I'll see if I can't address them or think of an alternative approach altogether to center the document title.

jeyip committed:
Shift document title 300px when sidebar is open

jeyip committed:
Add nav sidebar width the base style variables

jeyip committed:
Update references to nav sidebar width

jeyip commented on packages/base-styles/_variables.scss:
```diff
@@ -43,6 +43,7 @@ $button-size: 36px;
 $button-size-small: 24px;
 $header-height: 60px;
 $panel-header-height: $grid-unit-60;
+$nav-sidebar-width: 300px;
```
**Note:**

I'd prefer to scope this variable to the site editor package (in a new _variables.scss in `src/edit_site`), but it looks like the prevailing convention is to add variables to this global base-styles stylesheet. If anyone knows otherwise, I'm happy to make changes.

Addison-Stavlo commented:
Thinking more about this:
> If we have a scrollbar in the content then the document title is a little bit off. Since the editor's content container's width is smaller with a few pixels because of the scrollbar.

I don't think we should be too concerned with this side of things.  Consider that many users may often have the 'block-inspector' toolbar on the right hand side of the editor open a majority of the time, this shifts the editor canvas to the left by a the width of that sidebar.  We probably shouldn't be worried about it being perfectly centered over the editor canvas, but centered with respect to the topbar area itself?

jeyip commented on packages/edit-site/src/components/header/style.scss:
```diff
@@ -20,6 +20,13 @@
+// Keeps the document title centered when the sidebar is open
+.navigation-open {
+	.edit-site-header_center {
+		margin-left: $nav-sidebar-width;
```
**Note:**

Still seeing a little bit of wonkiness when we shrink the browser viewport with the sidebar open

jeyip replied:
It seems like the current solution doesn't play well when shrinking the viewport 🤔

noahtallen replied:
ah yeah. it looks fine when the sidebar is closed, but is a lot more sticky when it's open

jeyip replied:
Yeah currently mulling over the best way to approach this. I think we might have to make tradeoffs between absolute positioning, which causes [this](https://github.com/WordPress/gutenberg/pull/25630#pullrequestreview-496529493), and adding padding to the document title, which causes the "center div to be pushed over further than it should be when the distance between "left" and "right" is small enough".

noahtallen commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -88,7 +93,11 @@ export default function Header( {
-		<div className="edit-site-header">
+		<div
+			className={ classnames( 'edit-site-header', {
+				'navigation-open': isNavigationOpen,
```
should this go further up in the DOM? I feel like there is some precedent to set these at the `body` tag (e.g. `is-full-screen`)

jeyip replied:
This is a good point. Normally I'd say wait until there is a business use case, but I think we can safely predict that other components and styles will eventually need to leverage the same `is-navigation-open` flag.

Would you mind pointing me in the right direction? I'm having a little bit of trouble finding the `body` tag. If not, I'll go looking tomorrow.

noahtallen replied:
I just meant this: https://www.w3schools.com/tags/tag_body.asp. It's already where many similar styles like this exist :) I'm not sure where those styles are set; I'd probably search for `is-navigation-open` to see where it is added

jeyip replied:
> I just meant this: https://www.w3schools.com/tags/tag_body.asp.

Thanks Noah! Yeah, I was following the idea of the body tag. It was more of a question of where the styles are generated to add to the body tag, and I was wondering if it was anywhere obvious like an `index.html` in the Gutenberg repo.

I'll go searching now 👍

jeyip replied:
Hmmm...it looks like one of the patterns we use to establish `is-fullscreen-mode` is here `packages/interface/src/components/fullscreen-mode/index.js:36`

```javascript
if ( isActive ) {
	document.body.classList.add( 'is-fullscreen-mode' );
} else {
	document.body.classList.remove( 'is-fullscreen-mode' );
}
```

Based on a comment in the same file on line 11, the classname is also set in PHP somewhere when loading Gutenberg. I'll spend a little bit looking for the implementation, but I think we'll end up with a similar implementation as for fullscreen mode, where we set the classname with `document.body.classList.add()`

noahtallen replied:
Yep, that makes sense to me. I do not believe that the reference to PHP is applicable in this situation :) I do know that historically, classes like that would have been added in PHP for other admin screens or even the front end. It looks like that comment is just handling compatibility with a different flag which can cause issues

jeyip replied:
I added the body classname update in 13aad6b

jeyip commented:
> We probably shouldn't be worried about it being perfectly centered over the editor canvas in all cases, but centered with respect to the topbar area/items itself?

That's a good observation. I never noticed that opening the block settings sidebar changes content width. 

> many users may often have the 'block-inspector' toolbar on the right-hand side of the editor open a majority of the time

Given this information, addressing the scrollbar width seems less important. I do, however, think that we should fix the problem eventually. I imagine that there is still a large contingent of users that do not have block settings open when editing content. If they have a scrollbar permanently displayed, the document title will always be slightly misaligned.

How would we feel about creating a separate issue that's lower priority and handling the scrollbar in the future? @david-szabo97  @Addison-Stavlo

noahtallen reviewed (changes requested):
yeah, I think the margin causes the center div to be pushed over further than it should be when the distance between "left" and "right" is small enough. So this makes things messy at even medium-sized viewports.

I also noticed this issue:
<img width="1792" alt="Screen Shot 2020-09-28 at 3 40 35 PM" src="https://user-images.githubusercontent.com/6265975/94493318-24946700-01a1-11eb-8cab-2eb89e924325.png">

jeyip committed:
Use flex basis to center document title with open sidebar open

jeyip commented:
@noahtallen @Addison-Stavlo I updated the styling. The title should be centering correctly and all problems should be addressed ( (except the misalignment when a permanent scrollbar is present that @david-szabo97 called out)

david-szabo97 commented:
> Given this information, addressing the scrollbar width seems less important. I do, however, think that we should fix the problem eventually. I imagine that there is still a large contingent of users that do not have block settings open when editing content. If they have a scrollbar permanently displayed, the document title will always be slightly misaligned.
> How would we feel about creating a separate issue that's lower priority and handling the scrollbar in the future?

Yep! This is kinda out of scope for this PR. Let's do it separately.

noahtallen reviewed (approved):
other than the one thing about adding the classname to the body tag, I think this is about ready to go!

I'm noticing it still has two issues, but they already exist, so they don't have to block this.

The first is that the title looks off-center when the _right_ sidebars are open. I know it's still centered in the expected spot, but it just looks visually weird. I'm unsure what our expectations should be for this, so I think we can leave it be for now.
<img width="1907" alt="Screen Shot 2020-09-29 at 10 28 29 AM" src="https://user-images.githubusercontent.com/6265975/94594765-bb633100-023e-11eb-82c2-35de88342fd7.png">

This is probably a bug with the nav sidebar rather than this, but at smallish viewports, the size of the sidebar in the header is different from the size of the sidebar in the body 
<img width="1792" alt="Screen Shot 2020-09-28 at 3 40 35 PM" src="https://user-images.githubusercontent.com/6265975/94594749-b7371380-023e-11eb-9c8d-cb1dad8baccd.png">

jeyip commented:
> The first is that the title looks off-center when the right sidebars are open. I know it's still centered in the expected spot, but it just looks visually weird. I'm unsure what our expectations should be for this, so I think we can leave it be for now.

The same thing happens when the inserter sidebar is opened 🙃

jeyip committed:
Update body classname when navigation sidebar is open

jeyip commented:
Created an issue for the scrollbar misalignment @david-szabo97 https://github.com/WordPress/gutenberg/issues/25716

noahtallen commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -75,6 +76,14 @@ export default function Header( {
 		setPage,
 	} = useDispatch( 'core/edit-site' );
 
+	useEffect( () => {
```
This is a bit nitpicky, but I wonder if it would make sense to place this code closer to where the `isNavigationOpen` prop is set. Maybe in `editor/index.js`, could detect based on the `leftSidebarContent` variable

jeyip replied:
I don't have a strong opinion about this. I'll move it 🙂

noahtallen commented:
nice, the new body class is looking good! Left a nitpicky comment about code location, but this is ready to merge whenever you like ;) :shipit:

jeyip committed:
Set nav sidebar is open  body classname from editor index
